### PR TITLE
vscode-extensions: Add various C# extensions

### PR DESCRIPTION
**Pull request by magnouvean**
## Description of changes
Added 3 new vscode extensions, namely:
* [CSharpier](https://marketplace.visualstudio.com/items?itemName=csharpier.csharpier-vscode)
* [IntelliCode for C# Dev Kit](https://marketplace.visualstudio.com/items?itemName=ms-dotnettools.vscodeintellicode-csharp)
* [.NET Install Tool](https://marketplace.visualstudio.com/items?itemName=ms-dotnettools.vscode-dotnet-runtime)

The last of these doesn't make that much sense alone in `nixos` I don't think, but is a dependency of C# dev kit (and maybe the normal c# extension as well) so I think it makes sense to have it in here as these extensions will not even run without this extensions installed. The intellicode extension was the most work, but I did get it to work fine after a little while. The final package I did get working was heavily inspired by the `continue.continue` vscode-extension package declared in the same file. The only thing I am a little unsure about is whether `lib.optionals stdenv.isLinux` should be in the nativeBuildInputs (it was in `continue.continue`, but maybe it shouldn't be here).

I did test that all of the extensions work just as expected, but this was only tested using vscode on linux `x86_64` so outside this I am not sure.
I didn't add myself as a maintainer, but you can probably expect some maintenance prs in the future anyway, since I am planning on using C# quite a lot for quite some time :)

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/applications/editors/vscode/extensions/default.nix [hunk 1]
```diff
@@ -1099,6 +1099,18 @@ let
...
       };
 
+      csharpier.csharpier-vscode = buildVscodeMarketplaceExtension {
+        mktplcRef = {
+          name = "csharpier-vscode";
+          publisher = "csharpier";
+          version = "1.7.3";
+          hash = "sha256-/ZLjnlLl6xmgEazdCbnuE6UuuV1tDwAjpxz+vmBuYHE=";
+        };
+        meta = {
```
Comment (superherointj) [inline pkgs/applications/editors/vscode/extensions/default.nix:1109]: Meta is incomplete. Add changelog, downloadPage, description, homepage, maintainers, etc.
Consider making yourself a maintainer to be notified on updates.

File: pkgs/applications/editors/vscode/extensions/default.nix [hunk 2]
```diff
@@ -3109,6 +3121,62 @@ let
...
       ms-dotnettools.csharp = callPackage ./ms-dotnettools.csharp { };
 
+      ms-dotnettools.vscode-dotnet-runtime = buildVscodeMarketplaceExtension {
+        mktplcRef = {
+          name = "vscode-dotnet-runtime";
+          publisher = "ms-dotnettools";
+          version = "2.1.1";
+          hash = "sha256-k14bjWITPDduJi79W59SnMV2TFNRCeAymhs6u1Y0vzk=";
+        };
+        meta = {
```
Comment (superherointj) [inline pkgs/applications/editors/vscode/extensions/default.nix:3131]: Meta is incomplete.

File: pkgs/applications/editors/vscode/extensions/default.nix [hunk 3]
```diff
@@ -3109,6 +3121,62 @@ let
...
+            publisher = "ms-dotnettools";
+            version = "2.1.11";
+          }
+          // sources.${stdenv.system};
+        nativeBuildInputs = lib.optionals stdenv.isLinux [ autoPatchelfHook ];
+        buildInputs = [
+          stdenv.cc.cc.lib
+          zlib
+        ];
+        meta = {
```
Comment (superherointj) [inline pkgs/applications/editors/vscode/extensions/default.nix:3169]: Meta is incomplete

File: pkgs/applications/editors/vscode/extensions/default.nix [hunk 4]
```diff
@@ -3109,6 +3121,62 @@ let
...
+          license = lib.licenses.unfree;
+          platforms = [
```
Comment (superherointj) [inline pkgs/applications/editors/vscode/extensions/default.nix:3171]: > lib.platforms.unix

**Comment by superherointj** [comment]
The structure of commits is not right. Each package should be a single commit.
In this case, there should exist no nixfmt commit. Because you are adding a package.

File: pkgs/applications/editors/vscode/extensions/default.nix (hunk 4 above)
Comment (magnouvean) [reply pkgs/applications/editors/vscode/extensions/default.nix:3171]: Would that not include `i686` (32 bit), which isn't supported by the extension?

Comment (superherointj) [reply pkgs/applications/editors/vscode/extensions/default.nix:3171]: Most things just work. Still, don't worry. Keep as it is then.

File: pkgs/applications/editors/vscode/extensions/default.nix [hunk 5]
```diff
@@ -1099,6 +1099,23 @@ let
...
+          name = "csharpier-vscode";
+          publisher = "csharpier";
+          version = "1.7.3";
+          hash = "sha256-/ZLjnlLl6xmgEazdCbnuE6UuuV1tDwAjpxz+vmBuYHE=";
+        };
+        meta = {
+          changelog = "https://marketplace.visualstudio.com/items/csharpier.csharpier-vscode/changelog";
+          description = "CSharpier code formatter for Visual Studio Code";
+          downloadPage = "https://marketplace.visualstudio.com/items?itemName=csharpier.csharpier-vscode";
+          homePage = "https://github.com/belav/csharpier";
```
Comment (superherointj) [inline pkgs/applications/editors/vscode/extensions/default.nix:1113]:
```suggestion
          homepage = "https://github.com/belav/csharpier";
```
Error

File: pkgs/applications/editors/vscode/extensions/default.nix [hunk 6]
```diff
@@ -3109,6 +3126,72 @@ let
...
+        nativeBuildInputs = lib.optionals stdenv.isLinux [ autoPatchelfHook ];
+        buildInputs = [
+          stdenv.cc.cc.lib
+          zlib
+        ];
+        meta = {
+          changelog = "https://marketplace.visualstudio.com/items/ms-dotnettools.vscodeintellicode-csharp/changelog";
+          description = "AI-assisted development features for C# in Visual Studio Code";
+          downloadPage = "https://marketplace.visualstudio.com/items?itemName=ms-dotnettools.vscodeintellicode-csharp";
+          homePage = "https://github.com/MicrosoftDocs/intellicode";
```
Comment (superherointj) [inline pkgs/applications/editors/vscode/extensions/default.nix:3183]:
```suggestion
          homepage = "https://github.com/MicrosoftDocs/intellicode";
```

**Review by superherointj — APPROVED** [review]

**Pull request merged**